// @sanity/form-builder/lib/inputs/BlockEditor-slate/plugins/onPasteSlateContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slate = require("slate");

var _randomKey = _interopRequireDefault(require("../util/randomKey"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function formBuilderNodeOnPaste(formBuilder, editorFields) {
  function processNode(node) {
    if (!node.get('nodes')) {
      return node;
    }

    const newKey = (0, _randomKey.default)(12);
    const SlateType = node.constructor;
    const newData = node.get('data') ? node.get('data').toObject() : undefined;

    if (newData && newData.value && newData.value._key) {
      newData.value._key = newKey;
    }

    return new SlateType({
      data: _slate.Data.create(newData),
      isVoid: node.get('isVoid'),
      key: newKey,
      nodes: node.get('nodes').map(processNode),
      type: node.get('type')
    });
  }

  function onPaste(event, data, change) {
    if (typeof data.fragment === 'undefined' || data.fragment === null) {
      return null;
    }

    const newNodesList = _slate.Block.createList(data.fragment.nodes.toArray().map(processNode));

    const newDoc = new _slate.Document({
      key: data.fragment.key,
      nodes: newNodesList
    });
    return change.insertFragment(newDoc);
  }

  return {
    onPaste
  };
}

var _default = formBuilderNodeOnPaste;
exports.default = _default;
