// @sanity/components/lib/selects/FuzzySearchableSelect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _searchable = _interopRequireDefault(require("part:@sanity/components/selects/searchable"));

var _fuse = _interopRequireDefault(require("fuse.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class FuzzySearchableSelect extends _react.default.Component {
  constructor(props, context) {
    super(props, context);
    this.handleSearch = this.handleSearch.bind(this);
    const fuseOptions = {
      keys: ['title']
    };
    const items = this.props.items;
    this.fuse = new _fuse.default(items, fuseOptions);
    this.state = {
      searchResult: [],
      query: null
    };
  }

  handleSearch(query) {
    this.setState({
      searchResult: this.fuse.search(query)
    });
  }

  render() {
    const _this$props = this.props,
          label = _this$props.label,
          error = _this$props.error,
          value = _this$props.value,
          placeholder = _this$props.placeholder;
    const _this$state = this.state,
          searchResult = _this$state.searchResult,
          loading = _this$state.loading;
    return _react.default.createElement(_default.default, {
      error: error,
      label: label
    }, _react.default.createElement(_searchable.default, {
      placeholder: placeholder,
      value: value,
      onSearch: this.handleSearch,
      onChange: this.props.onChange,
      onFocus: this.props.onChange,
      loading: loading,
      items: searchResult
    }));
  }

}

exports.default = FuzzySearchableSelect;

_defineProperty(FuzzySearchableSelect, "propTypes", {
  label: _propTypes.default.string.isRequired,
  placeholder: _propTypes.default.string,
  onChange: _propTypes.default.func,
  value: _propTypes.default.object,
  error: _propTypes.default.bool,
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string
  }))
});

_defineProperty(FuzzySearchableSelect, "defaultProps", {
  onChange() {},

  onBlur() {},

  onFocus() {}

});
