// @sanity/components/lib/edititem/StoryMover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactDraggable = _interopRequireDefault(require("react-draggable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Mover extends _react.default.Component {
  constructor(props, args) {
    super(props, args);

    _defineProperty(this, "handleStop", (event, element) => {
      const scrollTop = document.getElementById('myScrollContainerId').scrollTop; // Trigger window resize to force redrawing of the popover component

      window.dispatchEvent(new Event('resize'));
      this.setState({
        x: event.pageX - event.offsetX,
        y: event.pageY - event.offsetY + scrollTop
      });
    });

    this.state = {
      x: 50,
      y: 50
    };
  }

  render() {
    return _react.default.createElement("div", {
      className: "moverElement"
    }, _react.default.createElement(_reactDraggable.default, {
      onStop: this.handleStop
    }, _react.default.createElement("div", {
      style: {
        zIndex: 1061,
        position: 'absolute',
        cursor: 'move',
        color: 'white',
        backgroundColor: 'red',
        padding: '10px'
      }
    }, "Move me")), _react.default.createElement("div", {
      style: {
        position: 'absolute',
        top: `${this.state.y + 50}px`,
        left: `${this.state.x - 10}px`
      }
    }, _react.default.createElement("div", {
      style: {
        color: 'red',
        position: 'absolute',
        zIndex: '5000',
        backgroundColor: 'black'
      }
    }, "x: ", this.state.x - 10, " y: ", this.state.y + 50), this.props.children));
  }

}

exports.default = Mover;

_defineProperty(Mover, "propTypes", {
  children: _propTypes.default.node
});
