// @sanity/components/lib/buttons/ButtonCollection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _buttonCollectionStyle = _interopRequireDefault(require("part:@sanity/components/buttons/button-collection-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ButtonCollection extends _react.default.PureComponent {
  render() {
    const _this$props = this.props,
          align = _this$props.align,
          children = _this$props.children,
          secondary = _this$props.secondary;
    return _react.default.createElement("div", {
      className: align === 'start' ? _buttonCollectionStyle.default.alignStart : _buttonCollectionStyle.default.alignEnd
    }, _react.default.createElement("div", {
      className: _buttonCollectionStyle.default.primary
    }, children), secondary && _react.default.createElement("div", {
      className: _buttonCollectionStyle.default.secondary
    }, secondary));
  }

}

exports.default = ButtonCollection;

_defineProperty(ButtonCollection, "propTypes", {
  children: _propTypes.default.node.isRequired,
  secondary: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.arrayOf([_propTypes.default.node])]),
  align: _propTypes.default.oneOf(['start', 'end'])
});

_defineProperty(ButtonCollection, "defaultProps", {
  align: 'start',
  secondary: null
});
