// @sanity/components/lib/toggles/Buttons.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _buttonsStyle = _interopRequireDefault(require("part:@sanity/components/toggles/buttons-style"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const ITEM_SHAPE = {
  icon: _propTypes.default.node,
  title: _propTypes.default.string
};

class ToggleButtons extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "handleClick", event => {
      const _this$props = this.props,
            onChange = _this$props.onChange,
            items = _this$props.items;
      const index = Number(event.currentTarget.getAttribute('data-index'));
      onChange(items[index]);
    });
  }

  render() {
    const _this$props2 = this.props,
          items = _this$props2.items,
          label = _this$props2.label,
          value = _this$props2.value;
    return _react.default.createElement("div", {
      className: _buttonsStyle.default.root
    }, _react.default.createElement("div", {
      className: _buttonsStyle.default.label
    }, label), items.map((item, i) => {
      return _react.default.createElement(_default.default, {
        className: item == value ? _buttonsStyle.default.selectedButton : _buttonsStyle.default.button,
        kind: "simple",
        key: i,
        icon: item.icon,
        onClick: this.handleClick,
        "data-index": i
      }, item.title);
    }));
  }

}

exports.default = ToggleButtons;

_defineProperty(ToggleButtons, "propTypes", {
  label: _propTypes.default.string,
  onChange: _propTypes.default.func,
  value: _propTypes.default.shape(ITEM_SHAPE),
  items: _propTypes.default.arrayOf(_propTypes.default.shape(ITEM_SHAPE))
});
