// @sanity/default-layout/lib/components/ActionModal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _router = require("part:@sanity/base/router");

var _default2 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _ActionModal = _interopRequireDefault(require("./styles/ActionModal.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ActionModal(props) {
  return _react.default.createElement(_default2.default, {
    className: _ActionModal.default.modal,
    title: props.title,
    onClose: props.onClose,
    isOpen: true
  }, _react.default.createElement("div", {
    className: _ActionModal.default.content
  }, _react.default.createElement("h1", null, "Create new"), _react.default.createElement("ul", {
    className: _ActionModal.default.list
  }, props.actions.map(action => {
    const Icon = action.icon;
    return _react.default.createElement("li", {
      className: _ActionModal.default.listItem,
      key: action.title
    }, _react.default.createElement(_router.IntentLink, {
      onClick: props.onClose,
      className: _ActionModal.default.actionLink,
      intent: "create",
      params: action.params
    }, _react.default.createElement("span", {
      className: _ActionModal.default.icon
    }, Icon ? _react.default.createElement(Icon, null) : _react.default.createElement(_fileIcon.default, null)), _react.default.createElement("span", null, action.title)));
  }))));
}

ActionModal.defaultProps = {
  title: 'Create',
  actions: []
};
ActionModal.propTypes = {
  title: _propTypes.default.string,
  actions: _propTypes.default.array,
  onClose: _propTypes.default.func.isRequired
};
var _default = ActionModal;
exports.default = _default;
