// @sanity/default-layout/lib/components/ToolSwitcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _router = require("part:@sanity/base/router");

var _ToolSwitcher = _interopRequireDefault(require("./styles/ToolSwitcher.css"));

var _pluginIcon = _interopRequireDefault(require("part:@sanity/base/plugin-icon"));

var _reactInk = _interopRequireDefault(require("react-ink"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToolSwitcher(props) {
  const tools = props.tools,
        router = props.router,
        activeToolName = props.activeToolName,
        onSwitchTool = props.onSwitchTool;
  return _react.default.createElement("div", {
    className: `${_ToolSwitcher.default.toolSwitcher} ${props.className}`
  }, _react.default.createElement("ul", {
    className: _ToolSwitcher.default.toolList
  }, tools.map(tool => {
    const itemClass = activeToolName === tool.name ? _ToolSwitcher.default.activeItem : _ToolSwitcher.default.item;
    const ToolIcon = tool.icon || _pluginIcon.default;
    const state = Object.assign({}, router.state, {
      tool: tool.name
    });
    return _react.default.createElement("li", {
      key: tool.name,
      className: itemClass
    }, _react.default.createElement(_router.StateLink, {
      className: _ToolSwitcher.default.toolLink,
      state: state,
      onClick: onSwitchTool
    }, _react.default.createElement("div", {
      className: _ToolSwitcher.default.iconContainer
    }, _react.default.createElement(ToolIcon, null)), _react.default.createElement("div", {
      className: _ToolSwitcher.default.toolName
    }, tool.title || tool.name), _react.default.createElement(_reactInk.default, {
      duration: 1000,
      opacity: 0.1,
      radius: 200
    })));
  })));
}

ToolSwitcher.defaultProps = {
  className: ''
};
ToolSwitcher.propTypes = {
  activeToolName: _propTypes.default.string,
  onSwitchTool: _propTypes.default.func,
  className: _propTypes.default.string,
  router: _propTypes.default.shape({
    state: _propTypes.default.object
  }),
  tools: _propTypes.default.arrayOf(_propTypes.default.shape({
    name: _propTypes.default.string.isRequired,
    icon: _propTypes.default.func
  }))
};

var _default = (0, _router.withRouterHOC)(ToolSwitcher);

exports.default = _default;
