// @sanity/form-builder/lib/inputs/NumberInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _PatchEvent = _interopRequireWildcard(require("../PatchEvent"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class NumberInput extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "_input", void 0);

    _defineProperty(this, "handleChange", event => {
      const nextValue = event.currentTarget.value;
      this.props.onChange(_PatchEvent.default.from(nextValue === '' ? (0, _PatchEvent.unset)() : (0, _PatchEvent.set)(Number(nextValue))));
    });

    _defineProperty(this, "setInput", input => {
      this._input = input;
    });
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    const _this$props = this.props,
          _this$props$value = _this$props.value,
          value = _this$props$value === void 0 ? '' : _this$props$value,
          readOnly = _this$props.readOnly,
          markers = _this$props.markers,
          type = _this$props.type,
          level = _this$props.level,
          onFocus = _this$props.onFocus;
    const validation = markers.filter(marker => marker.type === 'validation');
    const errors = validation.filter(marker => marker.level === 'error');
    return _react.default.createElement(_default2.default, {
      markers: markers,
      level: level,
      label: type.title,
      description: type.description
    }, _react.default.createElement(_default.default, {
      customValidity: errors && errors.length > 0 ? errors[0].item.message : '',
      type: "number",
      value: value,
      readOnly: readOnly,
      placeholder: type.placeholder,
      onChange: this.handleChange,
      onFocus: onFocus,
      ref: this.setInput
    }));
  }

}

exports.default = NumberInput;
