// @sanity/form-builder/lib/inputs/BlockEditor-slate/preview/ListItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ListItem = _interopRequireDefault(require("../styles/contentStyles/ListItem.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const listItems = ['bullet', 'number', 'roman'];

function ListItem(props) {
  const listItem = props.listItem;

  if (!listItems.includes(listItem)) {
    throw new Error(`Don't know how to handle listItem '${listItem}'. ` + `Expected one of '${listItems.join("', '")}'`);
  }

  const className = `${_ListItem.default[listItem]} ${_ListItem.default[`level-${props.level}`]}`;
  return _react.default.createElement("div", _extends({}, props.attributes, {
    className: className
  }), props.contentComponent({
    children: props.children
  }));
}

ListItem.propTypes = {
  attributes: _propTypes.default.object,
  listItem: _propTypes.default.oneOf(listItems),
  level: _propTypes.default.number,
  children: _propTypes.default.node,
  contentComponent: _propTypes.default.func
};
var _default = ListItem;
exports.default = _default;
