// get-it/lib/request/node-request.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var url = require('url');
var http = require('http');
var https = require('https');
var concat = require('simple-concat');
var follow = require('follow-redirects');
var timedOut = require('timed-out');
var isStream = require('is-stream');
var toStream = require('into-stream');
var objectAssign = require('object-assign');
var progressStream = require('progress-stream');
var decompressResponse = require('decompress-response');

var adapter = 'node';

// Reduce a fully fledged node-style response object to
// something that works in both browser and node environment
var reduceResponse = function reduceResponse(res, reqUrl, method, body) {
  return {
    body: body,
    url: reqUrl,
    method: method,
    headers: res.headers,
    statusCode: res.statusCode,
    statusMessage: res.statusMessage
  };
};

module.exports = function (context, cb) {
  var options = context.options;
  var uri = objectAssign({}, url.parse(options.url));
  var bodyType = isStream(options.body) ? 'stream' : _typeof(options.body);

  if (bodyType !== 'undefined' && bodyType !== 'stream' && bodyType !== 'string' && !Buffer.isBuffer(options.body)) {
    throw new Error('Request body must be a string, buffer or stream, got ' + bodyType);
  }

  var lengthHeader = {};
  if (options.bodySize) {
    lengthHeader['Content-Length'] = options.bodySize;
  } else if (options.body && Buffer.isBuffer(options.body)) {
    lengthHeader['Content-Length'] = options.body.length;
  }

  // Make sure callback is not called in the event of a cancellation
  var aborted = false;
  var callback = function callback(err, res) {
    return !aborted && cb(err, res);
  };
  context.channels.abort.subscribe(function () {
    aborted = true;
  });

  // Create a reduced subset of options meant for the http.request() method
  var reqOpts = objectAssign(uri, {
    method: options.method,
    headers: objectAssign({}, options.headers, lengthHeader)
  });

  // Allow middleware to inject a response, for instance in the case of caching or mocking
  var injectedResponse = context.applyMiddleware('interceptRequest', undefined, {
    adapter: adapter,
    context: context
  });

  // If middleware injected a response, treat it as we normally would and return it
  // Do note that the injected response has to be reduced to a cross-environment friendly response
  if (injectedResponse) {
    var cbTimer = setImmediate(callback, null, injectedResponse);
    var abort = function abort() {
      return clearImmediate(cbTimer);
    };
    return { abort: abort };
  }

  var protocol = uri.protocol === 'https:' ? https : http;

  // We're using the follow-redirects module to transparently follow redirects
  if (options.maxRedirects !== 0) {
    protocol = uri.protocol === 'https:' ? follow.https : follow.http;
    reqOpts.maxRedirects = options.maxRedirects || 5;
  }

  var request = protocol.request(reqOpts, function (response) {
    // See if we should try to decompress the response
    var tryDecompress = reqOpts.method !== 'HEAD';
    var res = tryDecompress ? decompressResponse(response) : response;

    var resStream = context.applyMiddleware('onHeaders', res, {
      headers: response.headers,
      adapter: adapter,
      context: context
    });

    // Concatenate the response body, then parse the response with middlewares
    concat(resStream, function (err, data) {
      if (err) {
        return callback(err);
      }

      var body = options.rawBody ? data : data.toString();
      var reduced = reduceResponse(res, response.responseUrl || options.url, // On redirects, `responseUrl` is set
      reqOpts.method, body);

      return callback(null, reduced);
    });
  });

  if (options.timeout) {
    timedOut(request, options.timeout);
  }

  request.once('error', callback);

  // Cheating a bit here; since we're not concerned about the "bundle size" in node,
  // and modifying the body stream would be sorta tricky, we're just always going
  // to put a progress stream in the middle here. Note that

  var _getProgressStream = getProgressStream(options),
      bodyStream = _getProgressStream.bodyStream,
      progress = _getProgressStream.progress;

  // Let middleware know we're about to do a request


  context.applyMiddleware('onRequest', { options: options, adapter: adapter, request: request, context: context, progress: progress });

  if (bodyStream) {
    bodyStream.pipe(request);
  } else {
    request.end(options.body);
  }

  return { abort: function abort() {
      return request.abort();
    } };
};

function getProgressStream(options) {
  if (!options.body) {
    return {};
  }

  var bodyIsStream = isStream(options.body);
  var length = options.bodySize || (bodyIsStream ? null : Buffer.byteLength(options.body));
  if (!length) {
    return bodyIsStream ? { bodyStream: options.body } : {};
  }

  var progress = progressStream({ time: 16, length: length });
  var bodyStream = bodyIsStream ? options.body : toStream(options.body);
  return { bodyStream: bodyStream.pipe(progress), progress: progress };
}
