// @sanity/schema/lib/core/traverseSchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = traverseSchema;
exports.UnknownType = void 0;

var _uniq2 = _interopRequireDefault(require("lodash/uniq"));

var _flatten2 = _interopRequireDefault(require("lodash/flatten"));

var _getDupes = require("../sanity/validation/utils/getDupes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const NOOP_VISITOR = typeDef => typeDef;

class UnknownType {
  constructor(name) {
    _defineProperty(this, "name", void 0);

    this.name = name;
  }

}

exports.UnknownType = UnknownType;
const TYPE_TYPE = {
  name: 'type',
  type: null
};
const FUTURE_RESERVED = ['any', 'time', 'date'];

function traverseSchema(types = [], coreTypes = [], visitor = NOOP_VISITOR) {
  const coreTypesRegistry = Object.create(null);
  const registry = Object.create(null);
  const coreTypeNames = coreTypes.map(typeDef => typeDef.name);
  const reservedTypeNames = FUTURE_RESERVED.concat(coreTypeNames);
  const typeNames = types.map(typeDef => typeDef.name).filter(Boolean);
  coreTypes.forEach(coreType => {
    coreTypesRegistry[coreType.name] = coreType;
  });
  types.forEach((type, i) => {
    // Allocate a placeholder for each type
    registry[type.name || `__unnamed_${i}`] = {};
  });

  function getType(typeName) {
    return typeName === 'type' ? TYPE_TYPE : coreTypesRegistry[typeName] || registry[typeName] || null;
  }

  const duplicateNames = (0, _uniq2.default)((0, _flatten2.default)((0, _getDupes.getDupes)(typeNames)));

  function isDuplicate(typeName) {
    return duplicateNames.includes(typeName);
  }

  function getTypeNames() {
    return typeNames.concat(coreTypeNames);
  }

  function isReserved(typeName) {
    return typeName === 'type' || reservedTypeNames.includes(typeName);
  }

  const visitType = isRoot => typeDef => {
    return visitor(typeDef, {
      visit: visitType(false),
      isRoot,
      getType,
      getTypeNames,
      isReserved,
      isDuplicate
    });
  };

  coreTypes.forEach(coreTypeDef => {
    Object.assign(coreTypesRegistry[coreTypeDef.name], visitType(coreTypeDef));
  });
  types.forEach((typeDef, i) => {
    Object.assign(registry[typeDef.name || `__unnamed_${i}`], visitType(true)(typeDef));
  });
  return {
    get(typeName) {
      const res = registry[typeName] || coreTypesRegistry[typeName];

      if (res) {
        return res;
      }

      throw new Error(`No such type: ${typeName}`);
    },

    has(typeName) {
      return typeName in registry || typeName in coreTypesRegistry;
    },

    getTypeNames() {
      return Object.keys(registry);
    },

    getTypes() {
      return this.getTypeNames().map(this.get);
    },

    toJSON() {
      return this.getTypes();
    }

  };
}
