// @sanity/components/lib/utilities/Poppable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactPopper = require("react-popper");

var _Poppable = _interopRequireDefault(require("./styles/Poppable.css"));

var _Stacked = _interopRequireDefault(require("./Stacked"));

var _Escapable = _interopRequireDefault(require("./Escapable"));

var _CaptureOutsideClicks = _interopRequireDefault(require("./CaptureOutsideClicks"));

var _Portal = require("./Portal");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Poppable extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "targetRef", _react.default.createRef());

    _defineProperty(this, "handleClickOutside", ev => {
      if (!this.targetRef.current.contains(ev.target)) {
        this.props.onClickOutside(ev);
      }
    });
  }

  render() {
    const _this$props = this.props,
          onEscape = _this$props.onEscape,
          placement = _this$props.placement,
          modifiers = _this$props.modifiers,
          onClickOutside = _this$props.onClickOutside,
          target = _this$props.target,
          children = _this$props.children;
    return _react.default.createElement(_reactPopper.Manager, null, _react.default.createElement(_reactPopper.Target, null, _react.default.createElement("div", {
      ref: this.targetRef
    }, target)), children && _react.default.createElement(_Portal.Portal, null, _react.default.createElement(_Stacked.default, null, isActive => _react.default.createElement("div", {
      className: _Poppable.default.portal
    }, _react.default.createElement(_reactPopper.Popper, {
      placement: placement,
      modifiers: modifiers
    }, _react.default.createElement(_Escapable.default, {
      onEscape: isActive ? onEscape : undefined
    }), onClickOutside ? _react.default.createElement(_CaptureOutsideClicks.default, {
      onClickOutside: isActive ? this.handleClickOutside : undefined
    }, children) : children)))));
  }

}

exports.default = Poppable;

_defineProperty(Poppable, "defaultProps", {
  placement: 'bottom-start',
  modifiers: {
    preventOverflow: 'viewport'
  }
});
