// @sanity/form-builder/lib/Preview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _PreviewAny = require("./utils/fallback-preview/PreviewAny");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Preview extends _react.default.Component {
  render() {
    const _this$props = this.props,
          type = _this$props.type,
          value = _this$props.value,
          layout = _this$props.layout;
    const PreviewComponent = this.context.formBuilder.resolvePreviewComponent(type);

    if (PreviewComponent) {
      return _react.default.createElement(PreviewComponent, {
        type: type,
        value: value,
        layout: layout
      });
    }

    return _react.default.createElement("div", {
      title: "Unable to resolve preview component. Using fallback."
    }, _react.default.createElement(_PreviewAny.PreviewAny, {
      value: value,
      maxDepth: 2
    }));
  }

}

exports.default = Preview;

_defineProperty(Preview, "propTypes", {
  layout: _propTypes.default.string,
  value: _propTypes.default.any,
  type: _propTypes.default.object.isRequired
});

_defineProperty(Preview, "contextTypes", {
  formBuilder: _propTypes.default.object
});
