// @sanity/core/lib/commands/cors/listCorsOriginsCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const helpText = `
Examples
  sanity cors list
`;
var _default = {
  name: 'list',
  group: 'cors',
  signature: '[ORIGIN]',
  helpText,
  description: 'List all origins allowed to access the API for this project',
  action: function () {
    var _action = _asyncToGenerator(function* (args, context) {
      const output = context.output;
      const apiClient = context.apiClient;
      const client = apiClient({
        requireUser: true,
        requireProject: true
      });
      const origins = yield client.request({
        url: '/cors'
      });
      output.print(origins.map(origin => origin.origin).join('\n'));
    });

    return function action(_x, _x2) {
      return _action.apply(this, arguments);
    };
  }()
};
exports.default = _default;
