// @sanity/util/lib/reduceConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _mergeWith2 = _interopRequireDefault(require("lodash/mergeWith"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const sanityEnv = process.env.SANITY_ENV || 'production';
const apiHosts = {
  staging: 'https://api.sanity.work',
  development: 'http://api.sanity.wtf'
};
const processEnvConfig = {
  project: process.env.STUDIO_BASEPATH ? {
    basePath: process.env.STUDIO_BASEPATH
  } : {}
};

function merge(objValue, srcValue, key) {
  if (Array.isArray(objValue)) {
    return objValue.concat(srcValue);
  } // Pass on to default merging strategy


  return undefined;
}

var _default = (rawConfig, env = 'development') => {
  const apiHost = apiHosts[sanityEnv];
  const sanityConf = apiHost ? {
    api: {
      apiHost
    }
  } : {};
  const envConfig = (rawConfig.env || {})[env] || {};
  const config = (0, _mergeWith2.default)({}, rawConfig, envConfig, sanityConf, processEnvConfig, merge);
  delete config.env;
  return config;
};

exports.default = _default;
