// @sanity/components/lib/loading/AppLoadingScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _spinnerIcon = _interopRequireDefault(require("part:@sanity/base/spinner-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const AppLoadingScreenStyles = `
@-webkit-keyframes sanity-app-loader__spin {
  0% {
    -webkit-transform: rotate(0deg);
            transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
            transform: rotate(360deg);
  }
}
@keyframes sanity-app-loader__spin {
  0% {
    -webkit-transform: rotate(0deg);
            transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
            transform: rotate(360deg);
  }
}
@-webkit-keyframes sanity-app-loading-screen__colorFade {
  0% {
    fill: #383838;
    opacity: 100%;
  }

  30% {
    fill: #383838;
    opacity: 50%;
  }

  60% {
    fill: #2097ac;
    opacity: 70%;
  }

  90% {
    fill: #f43;
    opacity: 10%;
  }
}

@keyframes sanity-app-loading-screen__colorFade {
  0% {
    fill: #383838;
    opacity: 100%;
  }

  30% {
    fill: #383838;
    opacity: 50%;
  }

  60% {
    fill: #2097ac;
    opacity: 70%;
  }

  90% {
    fill: #f43;
    opacity: 10%;
  }
}

.sanity-app-loading-screen__root {
  display: block;
}

.sanity-app-loading-screen__inner {
  position: fixed;
  top: 50vh;
  left: 50vw;
  -webkit-transform: translateX(-50%) translateY(-50%);
          transform: translateX(-50%) translateY(-50%);
  text-align: center;
}

.sanity-app-loading-screen__logo {
  font-size: 5em;
  color: #383838;
}

.sanity-app-loading-screen__logo .sanityIconAnimate path:nth-child(odd) {
  -webkit-animation-name: sanity-app-loading-screen__colorFade;
          animation-name: sanity-app-loading-screen__colorFade;
  -webkit-animation-duration: 5s;
          animation-duration: 5s;
  -webkit-animation-duration: 5s;
          animation-duration: 5s;
  -webkit-animation-iteration-count: infinite;
          animation-iteration-count: infinite;
  -webkit-animation-direction: alternate;
          animation-direction: alternate;
}

.sanity-app-loading-screen__logo .sanityIconAnimate path:nth-child(even) {
  -webkit-animation-name: sanity-app-loading-screen__colorFade;
          animation-name: sanity-app-loading-screen__colorFade;
  -webkit-animation-duration: 5s;
          animation-duration: 5s;
  -webkit-animation-duration: 5s;
          animation-duration: 5s;
  -webkit-animation-iteration-count: infinite;
          animation-iteration-count: infinite;
  -webkit-animation-direction: alternate-reverse;
          animation-direction: alternate-reverse;
}

.sanity-app-loading-screen__text {
  font-weight: 400
  font-size: 5em;
  font-family: sans-serif;
}

.sanity-app-loading-screen__spinner {
  display: block;
  z-index: 1;
  position: fixed;
  top: 75vh;
  left: 50vw;
  -webkit-transform: translateX(-50%);
          transform: translateX(-50%);
}
.sanity-app-loader__spinner-inner {
  position: absolute;
  height: 1em;
  width: 1em;
}

.sanity-app-loader__spinner-inner svg {
  -webkit-transform-origin: center center;
          transform-origin: center center;
  -webkit-animation-name: sanity-app-loader__spin;
          animation-name: sanity-app-loader__spin;
  -webkit-animation-duration: 2s;
          animation-duration: 2s;
  -webkit-animation-timing-function: linear;
          animation-timing-function: linear;
  -webkit-animation-iteration-count: infinite;
          animation-iteration-count: infinite;
}
`;

class AppLoadingScreen extends _react.default.Component {
  render() {
    return _react.default.createElement("div", {
      className: "sanity-app-loading-screen"
    }, _react.default.createElement("style", {
      type: "text/css"
    }, AppLoadingScreenStyles), _react.default.createElement("div", {
      className: "sanity-app-loading-screen__inner"
    }, _react.default.createElement("div", {
      className: "sanity-app-loading-screen__logo"
    }, _react.default.createElement(_sanityLogoIcon.default, null)), _react.default.createElement("div", {
      className: "sanity-app-loading-screen__text"
    }, this.props.text)), _react.default.createElement("div", {
      className: "sanity-app-loading-screen__spinner"
    }, _react.default.createElement("div", {
      className: "sanity-app-loader__spinner-inner"
    }, _react.default.createElement(_spinnerIcon.default, null))));
  }

}

exports.default = AppLoadingScreen;

_defineProperty(AppLoadingScreen, "propTypes", {
  text: _propTypes.default.string
});

_defineProperty(AppLoadingScreen, "defaultProps", {
  text: 'Restoring sanity…'
});
