// @sanity/components/lib/panes/SplitPaneWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * SplitController plucks the props from this component and distributes panes,
 * so while this component might seem useless, it has a purpose.
 * @todo See if this can be done in a more intuitive way without this extra layer
 */
class SplitPaneWrapper extends _react.default.PureComponent {
  /* eslint-disable react/no-unused-prop-types */
  render() {
    return this.props.children || _react.default.createElement("div", null);
  }

}

exports.default = SplitPaneWrapper;

_defineProperty(SplitPaneWrapper, "propTypes", {
  minWidth: _propTypes.default.number,
  maxWidth: _propTypes.default.number,
  defaultWidth: _propTypes.default.number,
  children: _propTypes.default.node
  /* eslint-enable react/no-unused-prop-types */

});

_defineProperty(SplitPaneWrapper, "defaultProps", {
  minWidth: 0,
  maxWidth: 0,
  defaultWidth: 0,
  children: undefined
});
