// @sanity/schema/lib/legacy/types/blocks/defaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_DECORATORS = exports.DECORATORS = exports.DECORATOR_STRIKE = exports.DECORATOR_UNDERLINE = exports.DECORATOR_CODE = exports.DECORATOR_EMPHASIS = exports.DECORATOR_STRONG = exports.DEFAULT_BLOCK_STYLES = exports.BLOCK_STYLES = exports.DEFAULT_LIST_TYPES = exports.LIST_TYPES = exports.DEFAULT_MARKS_FIELD = exports.DEFAULT_TEXT_FIELD = exports.DEFAULT_LINK_ANNOTATION = void 0;
const DEFAULT_LINK_ANNOTATION = {
  type: 'object',
  name: 'link',
  fields: [{
    name: 'href',
    type: 'url',
    title: 'Url'
  }]
};
exports.DEFAULT_LINK_ANNOTATION = DEFAULT_LINK_ANNOTATION;
const DEFAULT_TEXT_FIELD = {
  type: 'text',
  name: 'text',
  title: 'Text'
};
exports.DEFAULT_TEXT_FIELD = DEFAULT_TEXT_FIELD;
const DEFAULT_MARKS_FIELD = {
  name: 'marks',
  type: 'array',
  of: [{
    type: 'string'
  }],
  title: 'Marks'
};
exports.DEFAULT_MARKS_FIELD = DEFAULT_MARKS_FIELD;
const LIST_TYPES = {
  bullet: {
    title: 'Bullet',
    value: 'bullet'
  },
  numbered: {
    title: 'Numbered',
    value: 'number'
  }
};
exports.LIST_TYPES = LIST_TYPES;
const DEFAULT_LIST_TYPES = [LIST_TYPES.bullet, LIST_TYPES.numbered];
exports.DEFAULT_LIST_TYPES = DEFAULT_LIST_TYPES;
const BLOCK_STYLES = {
  normal: {
    title: 'Normal',
    value: 'normal'
  },
  h1: {
    title: 'Heading 1',
    value: 'h1'
  },
  h2: {
    title: 'H2',
    value: 'h2'
  },
  h3: {
    title: 'H3',
    value: 'h3'
  },
  h4: {
    title: 'H4',
    value: 'h4'
  },
  h5: {
    title: 'H5',
    value: 'h5'
  },
  h6: {
    title: 'H6',
    value: 'h6'
  },
  blockquote: {
    title: 'Quote',
    value: 'blockquote'
  }
};
exports.BLOCK_STYLES = BLOCK_STYLES;
const DEFAULT_BLOCK_STYLES = [BLOCK_STYLES.normal, BLOCK_STYLES.h1, BLOCK_STYLES.h2, BLOCK_STYLES.h3, BLOCK_STYLES.h4, BLOCK_STYLES.h5, BLOCK_STYLES.h6, BLOCK_STYLES.blockquote];
exports.DEFAULT_BLOCK_STYLES = DEFAULT_BLOCK_STYLES;
const DECORATOR_STRONG = {
  title: 'Strong',
  value: 'strong'
};
exports.DECORATOR_STRONG = DECORATOR_STRONG;
const DECORATOR_EMPHASIS = {
  title: 'Emphasis',
  value: 'em'
};
exports.DECORATOR_EMPHASIS = DECORATOR_EMPHASIS;
const DECORATOR_CODE = {
  title: 'Code',
  value: 'code'
};
exports.DECORATOR_CODE = DECORATOR_CODE;
const DECORATOR_UNDERLINE = {
  title: 'Underline',
  value: 'underline'
};
exports.DECORATOR_UNDERLINE = DECORATOR_UNDERLINE;
const DECORATOR_STRIKE = {
  title: 'Strike',
  value: 'strike-through'
};
exports.DECORATOR_STRIKE = DECORATOR_STRIKE;
const DECORATORS = {
  strong: DECORATOR_STRONG,
  em: DECORATOR_EMPHASIS,
  code: DECORATOR_CODE,
  underline: DECORATOR_UNDERLINE,
  strikeThrough: DECORATOR_STRIKE
};
exports.DECORATORS = DECORATORS;
const DEFAULT_DECORATORS = [DECORATORS.strong, DECORATORS.em, DECORATORS.code, DECORATORS.underline, DECORATORS.strikeThrough];
exports.DEFAULT_DECORATORS = DEFAULT_DECORATORS;
