// @sanity/server/lib/baseServer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBaseServer = getBaseServer;
exports.getDocumentElement = getDocumentElement;
exports.applyStaticRoutes = applyStaticRoutes;
exports.callInitializers = callInitializers;

var _path = _interopRequireDefault(require("path"));

var _express = _interopRequireDefault(require("express"));

var _react = _interopRequireDefault(require("react"));

var _server = _interopRequireDefault(require("react-dom/server"));

var _requireUncached = _interopRequireDefault(require("require-uncached"));

var _resolver = require("@sanity/resolver");

var _getStaticBasePath = _interopRequireDefault(require("./util/getStaticBasePath"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const docPart = 'part:@sanity/base/document';
const initPart = 'part:@sanity/server/initializer';

const getDefaultModule = mod => {
  return mod && mod.__esModule ? mod.default : mod;
};

const getTitle = (project = {}) => {
  const projectName = project && project.name || '';
  return projectName ? `${projectName} - Sanity` : 'Sanity';
};

const assetify = (assetPath, hashes) => ({
  path: assetPath,
  hash: hashes[assetPath]
});

const getDocumentComponent = basePath => (0, _resolver.resolveParts)({
  basePath
}).then(res => {
  const part = res.implementations[docPart];

  if (!part) {
    throw new Error(`Part '${docPart}' is not implemented by any plugins, are you missing @sanity/base?`);
  }

  return getDefaultModule((0, _requireUncached.default)(part[0].path));
});

function getBaseServer() {
  return (0, _express.default)();
}

function getDocumentElement({
  project,
  basePath,
  hashes
}, props = {}) {
  const assetHashes = hashes || {}; // Project filesystem base path

  return getDocumentComponent(basePath).then(Document => _react.default.createElement(Document, Object.assign({
    // URL base path
    basePath: process.env.STUDIO_BASEPATH || project && project.basePath || '',
    title: getTitle(project),
    stylesheets: ['css/main.css'].map(item => assetify(item, assetHashes)),
    scripts: ['js/vendor.bundle.js', 'js/app.bundle.js'].map(item => assetify(item, assetHashes))
  }, props)));
}

function applyStaticRoutes(app, config = {}) {
  const staticPath = config.staticPath || _path.default.join(__dirname, '..', 'public');

  const staticBasePath = (0, _getStaticBasePath.default)(config);
  app.use(staticBasePath, _express.default.static(staticPath));
  app.get('*', (req, res) => {
    if (req.url.startsWith(staticBasePath)) {
      return res.status(404).send('File not found');
    }

    return getDocumentElement(config).then(doc => res.send(`<!doctype html>${_server.default.renderToStaticMarkup(doc)}`)).catch(err => {
      console.error(err.stack); // eslint-disable-line no-console

      res.set('Content-Type', 'text/plain').status(500).send(err.stack);
    });
  });
  return app;
}

function callInitializers(config) {
  (0, _resolver.resolveParts)({
    config
  }).then(res => {
    const parts = res.implementations[initPart];

    if (!parts) {
      return;
    }

    res.implementations[initPart].map(part => getDefaultModule(require(part.path))).forEach(initializer => initializer(config));
  });
}
