// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/AnnotationButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _button = _interopRequireDefault(require("part:@sanity/components/toggles/button"));

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

var _AnnotationButton = _interopRequireDefault(require("./styles/AnnotationButton.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class AnnotationButton extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "handleToggleButtonClick", () => {
      this.props.onClick(this.props.annotation);
    });
  }

  render() {
    const annotation = this.props.annotation;
    return _react.default.createElement(_button.default, {
      onClick: this.handleToggleButtonClick,
      title: annotation.type.title,
      disabled: annotation.disabled,
      selected: annotation.active,
      className: _AnnotationButton.default.button,
      icon: _linkIcon.default
    });
  }

}

exports.default = AnnotationButton;

_defineProperty(AnnotationButton, "propTypes", {
  onClick: _propTypes.default.func,
  annotation: _propTypes.default.object
});
