// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/BlockStyle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.blockStyleShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _style = _interopRequireDefault(require("part:@sanity/components/selects/style"));

var _BlockStyle = _interopRequireDefault(require("./styles/BlockStyle.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const blockStyleShape = _propTypes.default.shape({
  key: _propTypes.default.string,
  active: _propTypes.default.bool,
  title: _propTypes.default.string,
  preview: _propTypes.default.node
});

exports.blockStyleShape = blockStyleShape;

class BlockStyle extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "renderItem", item => {
      return item.preview;
    });
  }

  render() {
    if (!this.props.items || this.props.items.length === 0) {
      return null;
    }

    return _react.default.createElement("label", {
      className: _BlockStyle.default.root
    }, _react.default.createElement("span", {
      style: {
        display: 'none'
      }
    }, "Text"), _react.default.createElement(_style.default, {
      items: this.props.items,
      value: this.props.value,
      onChange: this.props.onSelect,
      renderItem: this.renderItem,
      transparent: true
    }));
  }

}

exports.default = BlockStyle;

_defineProperty(BlockStyle, "propTypes", {
  value: _propTypes.default.arrayOf(blockStyleShape),
  items: _propTypes.default.arrayOf(blockStyleShape),
  onSelect: _propTypes.default.func
});
