// @sanity/form-builder/lib/inputs/BlockEditor-slate/toolbar/InsertBlocks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.insertBlockShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const insertBlockShape = _propTypes.default.shape({
  type: _propTypes.default.object,
  title: _propTypes.default.string
});

exports.insertBlockShape = insertBlockShape;

class InsertBlocks extends _react.default.Component {
  shouldComponentUpdate(nextProps, nextState) {
    return this.props.blocks !== nextProps.blocks;
  }

  render() {
    return _react.default.createElement(_dropdown.default, {
      items: this.props.blocks,
      onAction: this.props.onInsertBlock,
      kind: "simple",
      origin: "right"
    }, "Insert");
  }

}

exports.default = InsertBlocks;

_defineProperty(InsertBlocks, "propTypes", {
  onInsertBlock: _propTypes.default.func,
  blocks: _propTypes.default.arrayOf(insertBlockShape)
});
