// @sanity/components/lib/previews/BlockPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _blockStyle = _interopRequireDefault(require("part:@sanity/components/previews/block-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const fieldProp = _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.func]);

class BlockPreview extends _react.default.PureComponent {
  render() {
    const _this$props = this.props,
          title = _this$props.title,
          subtitle = _this$props.subtitle,
          description = _this$props.description,
          mediaDimensions = _this$props.mediaDimensions,
          media = _this$props.media,
          children = _this$props.children;
    return _react.default.createElement("div", {
      className: description ? _blockStyle.default.hasDescription : _blockStyle.default.root
    }, status && _react.default.createElement("div", {
      className: _blockStyle.default.status
    }, typeof status === 'function' && status({
      layout: 'default'
    }) || status), media && _react.default.createElement("div", {
      className: `${_blockStyle.default.media}`
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && _react.default.createElement("div", {
      className: _blockStyle.default.mediaString
    }, media), _react.default.isValidElement(media) && media), _react.default.createElement("div", {
      className: _blockStyle.default.heading
    }, _react.default.createElement("h2", {
      className: _blockStyle.default.title
    }, title), _react.default.createElement("h3", {
      className: _blockStyle.default.subtitle
    }, subtitle), description && _react.default.createElement("p", {
      className: _blockStyle.default.description
    }, description)), children && _react.default.createElement("div", {
      className: _blockStyle.default.children
    }, children));
  }

}

exports.default = BlockPreview;

_defineProperty(BlockPreview, "propTypes", {
  title: _propTypes.default.string,
  subtitle: _propTypes.default.string,
  description: _propTypes.default.string,
  mediaDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number,
    fit: _propTypes.default.oneOf(['clip', 'crop', 'fill', 'fillmax', 'max', 'scale', 'min']),
    aspect: _propTypes.default.number
  }),
  media: fieldProp,
  children: _propTypes.default.func,
  type: _propTypes.default.shape({
    title: _propTypes.default.string,
    name: _propTypes.default.string
  })
});

_defineProperty(BlockPreview, "defaultProps", {
  mediaDimensions: {
    width: 160,
    height: 160,
    aspect: 1,
    fit: 'crop'
  },
  type: {
    title: undefined,
    name: undefined
  }
});
