// @sanity/block-tools/lib/HtmlDeserializer/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRuleOptions = createRuleOptions;
exports.tagName = tagName;
exports.preprocess = preprocess;
exports.defaultParseHtml = defaultParseHtml;
exports.flattenNestedBlocks = flattenNestedBlocks;
exports.trimWhitespace = trimWhitespace;
exports.ensureRootIsBlocks = ensureRootIsBlocks;

var _constants = require("../constants");

var _blockContentTypeToOptions = _interopRequireDefault(require("../util/blockContentTypeToOptions"));

var _preprocessors = _interopRequireDefault(require("./preprocessors"));

var _resolveJsType = _interopRequireDefault(require("../util/resolveJsType"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * A utility function to create the options needed for the various rule sets,
 * based on the structure of the blockContentType
 *
 * @param {Object} The compiled schema type for the block content
 * @return {Object}
 */
function createRuleOptions(blockContentType) {
  const options = (0, _blockContentTypeToOptions.default)(blockContentType);
  const enabledBlockStyles = options.enabledBlockStyles || _constants.DEFAULT_SUPPORTED_STYLES;
  const enabledSpanDecorators = options.enabledSpanDecorators || _constants.DEFAULT_SUPPORTED_DECORATORS;
  const enabledBlockAnnotations = options.enabledBlockAnnotations || _constants.DEFAULT_SUPPORTED_ANNOTATIONS;
  return {
    enabledBlockStyles,
    enabledSpanDecorators,
    enabledBlockAnnotations
  };
}
/**
 * A utility function that always return a lowerCase version of the element.tagName
 *
 * @param {Object} DOMParser element
 * @return {String} Lowercase tagName for that element
 */


function tagName(el) {
  if (!el || el.nodeType !== 1) {
    return undefined;
  }

  return el.tagName.toLowerCase();
} // TODO: make this plugin-style


function preprocess(html, parseHtml, evaluate) {
  const compactHtml = html.trim() // Trim whitespace
  .replace(/\s\s+/g, ' ') // Remove multiple whitespace
  .replace(/[\r\n]/g, ' '); // Remove newlines / carriage returns

  const doc = parseHtml(compactHtml);

  _preprocessors.default.forEach(processor => {
    processor(html, doc, evaluate);
  });

  return doc;
}
/**
 * A default `parseHtml` function that returns the html using `DOMParser`.
 *
 * @param {String} html
 * @return {Object}
 */


function defaultParseHtml() {
  if ((0, _resolveJsType.default)(DOMParser) === 'undefined') {
    throw new Error('The native `DOMParser` global which the `Html` deserializer uses by ' + 'default is not present in this environment. ' + 'You must supply the `options.parseHtml` function instead.');
  }

  return html => {
    return new DOMParser().parseFromString(html, 'text/html');
  };
}

function flattenNestedBlocks(blocks) {
  let depth = 0;
  const flattened = [];

  const traverse = _nodes => {
    const toRemove = [];

    _nodes.forEach((node, i) => {
      if (depth === 0) {
        flattened.push(node);
      }

      if (node._type === 'block') {
        if (depth > 0) {
          toRemove.push(node);
          flattened.push(node);
        }

        depth++;
        traverse(node.children);
      }
    });

    toRemove.forEach(node => {
      _nodes.splice(_nodes.indexOf(node), 1);
    });
    depth--;
  };

  traverse(blocks);
  return flattened;
}

function trimWhitespace(blocks) {
  blocks.forEach(block => {
    const nextSpan = (child, index) => {
      const next = block.children[index + 1];
      return next && next._type === 'span' ? next : null;
    };

    const prevSpan = (child, index) => {
      const prev = block.children[index - 1];
      return prev && prev._type === 'span' ? prev : null;
    };

    block.children.forEach((child, index) => {
      if (child._type !== 'span') {
        return;
      }

      const nextChild = nextSpan(child, index);
      const prevChild = prevSpan(child, index);

      if (index === 0) {
        child.text = child.text.replace(/^[^\S\n]+/g, '');
      }

      if (index === block.children.length - 1) {
        child.text = child.text.replace(/[^\S\n]+$/g, '');
      }

      if (/\s/.test(child.text.substring(child.text.length - 1)) && nextChild && /\s/.test(nextChild.text.substring(0, 1))) {
        child.text = child.text.replace(/[^\S\n]+$/g, '');
      }

      if (/\s/.test(child.text.substring(0, 1)) && prevChild && /\s/.test(prevChild.text.substring(prevChild.text.length - 1))) {
        child.text = child.text.replace(/^[^\S\n]+/g, '');
      }

      if (!child.text) {
        block.children.splice(index, 1);
      }
    });
  });
  return blocks;
}

function ensureRootIsBlocks(blocks) {
  return blocks.reduce((memo, node, i, original) => {
    if (node._type === 'block') {
      memo.push(node);
      return memo;
    }

    if (i > 0 && original[i - 1]._type !== 'block') {
      const block = memo[memo.length - 1];
      block.children.push(node);
      return memo;
    }

    const block = _objectSpread({}, _constants.DEFAULT_BLOCK, {
      children: [node]
    });

    memo.push(block);
    return memo;
  }, []);
}
